### PR TITLE
Html

### PR DESCRIPTION
**Pull request by supersquirel500**
feat: added tag parsing
from tags, constructs complete forms and returns complete forms as a list.

### CODE REVIEW

**Comment by supersquirel500** [comment]
update:
1) reset local repo to match current upstream and repushed changes for html.py to origin.
2) got rid of C Macro like variables. ( decrease in readability, slight increase in memory footprint
3) removed computer generated file.

**Comment by supersquirel500** [comment]
update:
1) added collect() after form parse to remove old variables.

File: uPy/html.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    if(html == ''):
+        return []
+    
+    # Note: Tags can end on same line they were created on
+
+    # State List
+    # NOT_IN_TAG = 0  # 0 = not in a tag. b'<' not seen yet
+    # IN_TAG = 1      # 1 = inside a tag. b'<' seen but b'>' not seen
+
+    state = 0
```
Comment (comptonben) [inline uPy/html.py]: change this to "in_tag = false" and delete state altogether as well as comments above

File: uPy/html.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
+# Simple HTML parser for PMT to use
```
Comment (comptonben) [inline uPy/html.py]: Remove your name and add the PMT header from another file then update to what this file needs in that header.  Everything below that is fine, except the structure of return value, delete that you are reiterating

File: uPy/html.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+# Curran Hyde
```
Comment (comptonben) [inline uPy/html.py]: Remove your name and add the PMT header from another file then update to what this file needs in that header. Everything below that is fine, except the structure of return value, delete that you are reiterating

File: uPy/html.py (hunk 1 above)
Comment (comptonben) [inline uPy/html.py]: change this to "in_tag = false" and delete state altogether as well as comments above

File: uPy/html.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    # IN_TAG = 1      # 1 = inside a tag. b'<' seen but b'>' not seen
+
+    state = 0
+    i = 0
+    print("Entering Form parse loop")
+    while i < len(html):
+        byte = html[i]
+
+        # STATE 0 searching ..
+        if(state==0):
```
Comment (comptonben) [inline uPy/html.py]: if in_tag: (remove parenthesis)

File: uPy/html.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            if(byte != '<' ):
+                i=i+1
+                continue
+            else:
+                state=1
+                #print("Tag Detected: <")
+                
+        # STATE 1 in a tag
+        if(state==1):
```
Comment (comptonben) [inline uPy/html.py]: remove parenthesis, if in_tag

File: uPy/html.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    # NOT_IN_TAG = 0  # 0 = not in a tag. b'<' not seen yet
+    # IN_TAG = 1      # 1 = inside a tag. b'<' seen but b'>' not seen
+
+    state = 0
+    i = 0
+    print("Entering Form parse loop")
+    while i < len(html):
+        byte = html[i]
+
+        # STATE 0 searching ..
```
Comment (comptonben) [inline uPy/html.py]: make this comment something without state e.g., # parsing for beginning of tag

File: uPy/html.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        if(state==0):
+            if(byte != '<' ):
+                i=i+1
+                continue
+            else:
+                state=1
```
Comment (comptonben) [inline uPy/html.py]: change this to in_tag

File: uPy/html.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+           # '>' not found in html. tag starts but does not terminate
+            if(endpos==-1):
+                print("\tError: Tag does not terminate. No >")
+                #print(html[i:len(html)]) #print out the rest of the html for debugging
+                return []
+            
+            # complete tag found '<' and '>'
+            else:
+                # pull out '<' through '>'
+                tag=html[ i:endpos+1 ]
```
Comment (comptonben) [inline uPy/html.py]: this is nitpicky but remove the spaces before and after the [ ] and choose whether you want to have spaces before and after every '=' or not spaces, you change it up all over and sometimes mix and match

File: uPy/html.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                #print out the complete tag < ... >
+                #print("tag={0}".format(tag))
+                tag=tag.lstrip('<')
+                tag=tag.rstrip('>')
+                tag=tag.strip() #beg end whitespace
+                
+                #time to split up elements within the tag
+                #cannot just split by spaces values can have spaces between quotes
+                j=0
+                tag_state = 0
```
Comment (comptonben) [inline uPy/html.py]: make this a bool

File: uPy/html.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                tag_list =[]
+                while len(tag)>0:
+                    
+                    # searching for type
+                    if(tag_state == 0):
```
Comment (comptonben) [inline uPy/html.py]: if not tag_state and remove parenthesis

File: uPy/html.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+                            ret=tag
+                        else:
+                            ret=tag[0:j]
+                        
+                        tag=tag[j+1:len(tag)]
+                        ret=ret.strip()
+                        #print("type={0}\ttag={1}".format(ret,tag))
+                        tag_list.append(ret)
+                        
+                        tag_state=1
```
Comment (comptonben) [inline uPy/html.py]: make bool

File: uPy/html.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+                    
+                    #searching for key/value pair
+                    elif(tag_state == 1):
```
Comment (comptonben) [inline uPy/html.py]: bool and remove parenthesis

File: uPy/html.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+                    
+                        #key found, pull "value"
+                        #pull val inbetween quotes
+                        j = tag.find('"', 0, len(tag))
+                        tag=tag[j+1:len(tag)]
+                        j = tag.find('"', 0, len(tag))
+                        val=tag[0:j]
+                        tag = tag[j+1:len(tag)]
+                        val = val.strip()
+                        print("val={0}\ttag={1}".format(val,tag))
```
Comment (comptonben) [inline uPy/html.py]: remove this print

File: uPy/html.py (hunk 3 above)
Comment (supersquirel500) [reply uPy/html.py]: reqst.py doesn't follow this convention

File: uPy/html.py (hunk 4 above)
Comment (supersquirel500) [reply uPy/html.py]: removed state component altogether

File: uPy/html.py (hunk 5 above)
Comment (supersquirel500) [reply uPy/html.py]: removed state component altogether

File: uPy/html.py (hunk 6 above)
Comment (supersquirel500) [reply uPy/html.py]: removed states altogether

File: uPy/html.py (hunk 9 above)
Comment (supersquirel500) [reply uPy/html.py]: removed states altogether

File: uPy/html.py (hunk 10 above)
Comment (supersquirel500) [reply uPy/html.py]: removed states altogether

**Comment by supersquirel500** [comment]
> 
> 
> What's the Worse Case execution time required to parse a page of 50000 bytes? Big O?

Intentionally O(n).

Two requirements I set for this code were:
1) No recursion
2) No looping back over previous information. Only move forward.

File: uPy/html.py (hunk 3 above)
Comment (comptonben) [reply uPy/html.py]: Go ahead and fix reqst too then, I must have missed it on the initial PR

File: uPy/html.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+#                   dict["method"] -> "POST"
+def get_forms(html):
+    
+    html = html.decode('utf-8') # int to str
+
+    # each entry in list is "<form ...> <...> </form>" 
+    forms = get_tags(html, "form")
+    if(len(forms) == 0):
+        return []
+    elif(len(forms) == 1): #caused if <form...> but bad html received no </form>
```
Comment (comptonben) [inline uPy/html.py:197]: I still do not think we should be trying to fix splashpages bad html, we should assume its written correctly and anyone who has bad html is in the 50% we don't break, I'll let you ask the others what they think.

Comment (comptonben) [reply uPy/html.py:197]: This also is a very bad conditional...most pages only have 1 form so checking for this does not mean that there was an incomplete form

File: uPy/html.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+        forms = []      #get rid of previous bad entry
+        ret_val = find_complete_string(html, "<form", "</div", 0) #WORKAROUND: try going until the end of the next div
+        if(ret_val[0] == ""):
+            return []   #gah lee you've just hit the worst written splashpage ever. just give up already
+        forms.append(ret_val[0])
+    
+    #debugging
+    print("All forms as complete strings:")
```
Comment (comptonben) [inline uPy/html.py:205]: now that we're out of testing this, lets remove debugging statements that don't have meaning to us

File: uPy/html.py [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+        i+=1
+
+
+    #break forms as strings into a list of tags: [form] = ["<form ...>", "<...>", "</form>"]
+    all_forms = []
+    for string in forms:
+        print("Before pulling out tags: {0}".format(string))
+        ret = get_tags(string)
+        all_forms.append(ret)
+        print("After pulling out tags: {0}".format(ret))
```
Comment (comptonben) [inline uPy/html.py:218]: do these before and after print statements need to be here? based on the tag printed above we should know how the data is structured at this point if it works

File: uPy/html.py [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+
+    #debugging
+    print("\nEach form broken into a list of strings:")
```
Comment (comptonben) [inline uPy/html.py:221]: remove debugging statement

File: uPy/html.py [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+    i=0
+    while i < len(all_forms):
+        print( "form[{0}]=".format(i) )
```
Comment (comptonben) [inline uPy/html.py:224]: we are printing so much stuff, remember every single print creates a string object that is dangling in memory, what if the whole page is a big form? do we have enough room to print 20,000 bites 3 times and have all of it in memory to begin with as variables?

File: uPy/html.py [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+        for tag in all_forms[i]:
+            print(tag)
```
Comment (comptonben) [inline uPy/html.py:226]: printing again

File: uPy/html.py [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+        i+=1
+
+    # split out each tag's internals
+    print("\nSplitting up tag internals:")
```
Comment (comptonben) [inline uPy/html.py:230]: remove debug prints

File: uPy/html.py [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+    i = 0
+    while i < len(all_forms):
+        tags = all_forms[i]
+        j = 0
+        while j < len(tags):
+            tag = tags[j]
+            print("Tag before breakup: {0}".format(tag))
+            all_forms[i][j] = breakup_tag(tag)
+            print("After Breakup Form[{0}] tag[{1}]: {2}".format(i, j, all_forms[i][j]) )
```
Comment (comptonben) [inline uPy/html.py:239]: before and after tag prints, just comment the structure of the data no need for these prints

File: uPy/html.py [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+            j+=1
+        i+=1
+    
+
+    parsed_forms = []
+    for form in all_forms:
+        parsed_forms.append(construct_form_from_tag_internals(form))
+        print("Complete forms as structures:")
```
Comment (comptonben) [inline uPy/html.py:247]: remove these prints

File: uPy/wifi_connect.py [hunk 23]
```diff
@@ -79,6 +80,12 @@ def station_connected(station: WLAN, host: String, gdt: GDT, wifiLogger: Logger)
             print("Splashpage Length [{}]".format(len(splashpage)))
             
             print("Splashpage Breaking...")
+
+            # break with form resubmission
+            forms = get_forms(splashpage)
```
Comment (comptonben) [inline uPy/wifi_connect.py:85]: can you comment these lines out like with the a tag breaking method so that we're not automatically using this

File: uPy/html.py [hunk 24]
```diff
@@ -0,0 +1,255 @@
...
+        print(parsed_forms)
+    
+    return parsed_forms
+
+
+
+def get_internal_form_data(form):
```
Comment (comptonben) [inline uPy/html.py:254]: is this function used? if not delete it

File: uPy/html.py [hunk 25]
```diff
@@ -0,0 +1,255 @@
...
+                #print("\ttagDict[{0}]={1}".format(key,val))
+            
+            # other tags will be inside forms
+            tagDict["inside"] = []
+            form = tagDict
+            formOpen = True
+
+        # /form tag : end of form
+        elif(tag_internals[0]=="/form"):
+            #print("\ttag_type=/form")
```
Comment (comptonben) [inline uPy/html.py:152]: remove print here

File: uPy/html.py [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+            #print("tagDict[tag_type]= form")
+
+            # insert key="value" pairs
+            j=1
+            while(j<len(tag_internals)):
+                key=tag_internals[j]
+                val=tag_internals[j+1]
+                j+=2
+                tagDict[key] = val
+                #print("\ttagDict[{0}]={1}".format(key,val))
```
Comment (comptonben) [inline uPy/html.py:143]: remove print here

File: uPy/html.py [hunk 27]
```diff
@@ -0,0 +1,255 @@
...
+
+    for tag_internals in tags: # the splitup tag internals of a single tag
+
+        #form tag: beginning of form
+        if(tag_internals[0]=="form"):
+
+            # Create a new dictionary to store info
+            tagDict = {}
+            tagDict["tag_type"]=tag_internals[0]
+            #print("tagDict[tag_type]= form")
```
Comment (comptonben) [inline uPy/html.py:134]: remove print here

File: uPy/html.py [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+        #print("val={0}\ttag={1}".format(val,tag))
+        contents_list.append(val)
+
+    return contents_list
+
+
+
+def construct_form_from_tag_internals(tags):
+    
+    formOpen = False #indicates if <form ...> seen but </form> not seen yet
```
Comment (comptonben) [inline uPy/html.py:121]: change this to form_open you can search and replace all

File: uPy/html.py [hunk 29]
```diff
@@ -0,0 +1,255 @@
...
+    #From here on we have a list of the contents of a tag.
+    #Decide how to handle it
+
+    for tag_internals in tags: # the splitup tag internals of a single tag
+
+        #form tag: beginning of form
+        if(tag_internals[0]=="form"):
+
+            # Create a new dictionary to store info
+            tagDict = {}
```
Comment (comptonben) [inline uPy/html.py:132]: change this to tag_dict, search all refactor again

File: uPy/html.py [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+        forms = []      #get rid of previous bad entry
+        ret_val = find_complete_string(html, "<form", "</div", 0) #WORKAROUND: try going until the end of the next div
+        if(ret_val[0] == ""):
+            return []   #gah lee you've just hit the worst written splashpage ever. just give up already
+        forms.append(ret_val[0])
+    
+    #debugging
+    print("All forms as complete strings:")
+    i=0
+    while i < len(forms):
```
Comment (comptonben) [inline uPy/html.py:207]: is this print helpful, if so, it should be a for loop:

`for i, form in enumerate(forms):
     print("...".format(i, form))`

File: uPy/html.py [hunk 31]
```diff
@@ -0,0 +1,255 @@
...
+    while i < len(all_forms):
+        print( "form[{0}]=".format(i) )
+        for tag in all_forms[i]:
+            print(tag)
+        i+=1
+
+    # split out each tag's internals
+    print("\nSplitting up tag internals:")
+    i = 0
+    while i < len(all_forms):
```
Comment (comptonben) [inline uPy/html.py:232]: these two while loops:

for tags in all_forms:
    for tag in tags:
          ...

Comment (comptonben) [reply uPy/html.py:232]: none of the indenting is showing up so let me know if you don't understand

File: uPy/html.py [hunk 32]
```diff
@@ -0,0 +1,255 @@
...
+    else:       #ex: <form method="POST" ... >
+        ret=tag[0:j]
+    
+    tag=tag[j+1:len(tag)]
+    ret=ret.strip()
+    #print("type={0}\ttag={1}".format(ret,tag))
+    contents_list.append(ret)
+        
+    while len(tag)>0:
+        #print()
```
Comment (comptonben) [inline uPy/html.py:92]: remove this print

File: uPy/html.py [hunk 33]
```diff
@@ -0,0 +1,255 @@
...
+    # pull out type
+    j=tag.find(' ', 0, len(tag))
+    if(j == -1): #ex: </form>
+        ret=tag
+    else:       #ex: <form method="POST" ... >
+        ret=tag[0:j]
+    
+    tag=tag[j+1:len(tag)]
+    ret=ret.strip()
+    #print("type={0}\ttag={1}".format(ret,tag))
```
Comment (comptonben) [inline uPy/html.py:88]: remove print

File: uPy/html.py [hunk 34]
```diff
@@ -0,0 +1,255 @@
...
+
+        #searching for key/value pair
+        j=tag.find('=', 0, len(tag))
+        if(j==-1):
+            break #end if no keyval pairs
+
+        key = tag[0:j] #pull out key
+        tag = tag[j+1:len(tag)]
+        key = key.strip(' ')
+        #print("key={0}\ttag={1}".format(key,tag))
```
Comment (comptonben) [inline uPy/html.py:102]: remove print

File: uPy/html.py [hunk 35]
```diff
@@ -0,0 +1,255 @@
...
+        contents_list.append(key)
+    
+        #key found, pull "value"
+        #pull val inbetween quotes
+        j = tag.find('"', 0, len(tag)) #first quote
+        tag=tag[j+1:len(tag)]
+        j = tag.find('"', 0, len(tag)) #second quote
+        val=tag[0:j].strip()
+        tag = tag[j+1:len(tag)]
+        #print("val={0}\ttag={1}".format(val,tag))
```
Comment (comptonben) [inline uPy/html.py:112]: remove print

File: uPy/html.py [hunk 36]
```diff
@@ -0,0 +1,255 @@
...
+    
+    return tag_list
+
+
+
+def breakup_tag(tag):
+    tag=tag.lstrip('<')
+    tag=tag.rstrip('>')
+    tag=tag.strip() #beg end whitespace
+    #print("tag={0}".format(tag)) #print out the complete tag < ... >
```
Comment (comptonben) [inline uPy/html.py:72]: remove print

File: uPy/html.py [hunk 37]
```diff
@@ -0,0 +1,255 @@
...
+    j=0
+    ret_val = ["_",i,j]
+    
+    # look for a specific tag
+    if tag_type is not None:
+        #Sprint("DEBUG: Finding specific tag")
+        while ret_val[0] != "":
+            # tags that don't make up objects. ex: <form ...> ... </form>
+            ret_val = find_complete_string(html_as_string, "<{0}".format(tag_type), "</{0}>".format(tag_type), ret_val[2] ) #start search at end of last string
+            #print("DEBUG: object search attempt= {0}".format(ret_val))
```
Comment (comptonben) [inline uPy/html.py:44]: remove print

File: uPy/html.py [hunk 38]
```diff
@@ -0,0 +1,255 @@
...
+
+            # tags that do make up objects. ex: <a ... > there is no </a>
+            # start string found
+            if(ret_val[1] != -1):
+                if(ret_val[2] == -1): # end string was not found
+                    ret_val = find_complete_string(html_as_string, "<{0}".format(tag_type), ">", ret_val[1] )
+                    tag_list.append(ret_val[0])
+                    #print("DEBUG: tag search attempt= {0}".format(ret_val))
```
Comment (comptonben) [inline uPy/html.py:52]: remove print

File: uPy/html.py [hunk 39]
```diff
@@ -0,0 +1,255 @@
...
+                else: # end string was found
+                    tag_list.append(ret_val[0])
+
+    # look for all tags
+    else:
+        #print("DEBUG: Finding all tags")
```
Comment (comptonben) [inline uPy/html.py:58]: remove print

File: uPy/html.py (hunk 14 above)
Comment (supersquirel500) [reply uPy/html.py:197]: It does mean there was an incomplete form. This variable is a list of all tags from <form ...> through </form> inclusive.
If the length is zero, then neither <form ...> or </form> were found.
If the length is 1 then <form ...> was found but </form> was not so the find all tags function automatically catches this and ends the tag on > instead of </form>, This works for any tag of the format <x ...> </x>. It is best to not generate an error but rather catch smoothly and recover in a way that can be checked.

tldr, the code is correct as it represents the number of tags from <x ... > to </x> inclusive. So the length from <form ... > without a </form> is just the starting <form ...>, which is one.

File: uPy/html.py (hunk 15 above)
Comment (supersquirel500) [reply uPy/html.py:205]: gotcha

File: uPy/html.py [hunk 40]
```diff
@@ -0,0 +1,215 @@
...
+
+def get_tags(html_as_string, tag_type=None):
+    tag_list = []
+    i=0
+    j=0
+    ret_val = ["_",i,j]
+    
+    # look for a specific tag
+    if tag_type is not None:
+        #Sprint("DEBUG: Finding specific tag")
```
Comment (comptonben) [inline uPy/html.py:40]: remove print

File: uPy/html.py (hunk 14 above)
Comment (comptonben) [reply uPy/html.py:184]: this is good
